### PR TITLE
Fix install of symlinks with installopt but without installoptopt

### PR DESCRIPTION
**Pull request by whitequark**
Since the `ocamlc -> ocamlc.byte` symlink is not installed when ocamlopt exists, building ocamlopt, but not *.opt compilers, results in no ocamlc symlink being installed at all. Since ln -sf is already used everywhere, just make the code less clever, and let the symlink be overwritten when appropriate.

### CODE REVIEW

**Comment by whitequark** [comment]
(I suppose the reason no one hit this before is that the chief reason to have ocamlopt but not ocamlopt.opt is cross-compilation...)

**Comment by gasche** [comment]
I'm not fond of the fact that your solution implies that the order in which you run the `install` rules impacts the result. It will be fine if you just `make install`, but maybe during development people run more specific rules manually and now they have to be careful about that.

What about "make the test more clever"? It seems to me that the code previously had a kind of "if `ocamlopt` exists then make `installopt`" logic, on which the more clever linking was piggy-backed. We could just separate the two things, make `installopt` when `ocamlopt` exist, but install `ocamlc.byte` as `ocamlc` when `ocamlc.opt` does not exist. That seems simple (not "too clever for its own good"), and correct.

The specification I would expect is the following one:
- install commands reason on the state of what has been built
- they install the "short name" symlink exactly when they detect that the version they are installing is the best that is available

It is not followed today, but it does not sound hard to follow it. Or maybe this rule is too simplistic for cross-compilation scenarios?

**Comment by whitequark** [comment]
> I'm not fond of the fact that your solution implies that the order in which you run the install rules impacts the result. It will be fine if you just make install, but maybe during development people run more specific rules manually and now they have to be careful about that.

Unless I'm missing something, this is not the case, because `install` implies `installopt` implies `installoptopt`, whereas `installoptopt` installs more specific symlinks than `installopt` than `install`.

That said, your solution does seem more elegant to me also, so I can do that.

**Comment by whitequark** [comment]
@gasche Done. Remarkably the patch even became simpler.

File: Makefile [hunk 1]
```diff
@@ -275,7 +275,11 @@ installopt:
 	for i in $(OTHERLIBRARIES); \
 	  do (cd otherlibs/$$i; $(MAKE) installopt) || exit $$?; done
 	if test -f ocamlopt.opt ; then $(MAKE) installoptopt; else \
-	   cd $(INSTALL_BINDIR); ln -sf ocamlopt.byte$(EXE) ocamlopt$(EXE); fi
+	   cd $(INSTALL_BINDIR); \
+	   ln -sf ocamlc.byte$(EXE) ocamlc$(EXE); \
+	   ln -sf ocamlopt.byte$(EXE) ocamlopt$(EXE); \
+	   ln -sf ocamllex.byte$(EXE) ocamllex$(EXE); \
+	   fi
```
Comment (dra27) [inline Makefile]: Indentation is wrong for the `fi`

Comment (whitequark) [reply Makefile]: Fixed

File: Makefile [hunk 2]
```diff
@@ -254,7 +254,7 @@ install:
 	   cd $(INSTALL_BINDIR); \
 	   ln -sf ocamlc.byte$(EXE) ocamlc$(EXE); \
 	   ln -sf ocamllex.byte$(EXE) ocamllex$(EXE); \
-	   fi
+	fi
```
Comment (dra27) [inline Makefile:257]: FWIW The indentation of this one was correct, it's the three lines above which should be indented :smile:

Comment (dra27) [inline Makefile]: FWIW The indentation of this one was correct, it's the three lines above which should be indented :smile:

**Comment by dra27** [comment]
@gasche - is a Changes entry needed for this, in your opinion?

File: Makefile (hunk 2 above)
Comment (whitequark) [reply Makefile:257]: This makes no sense, this conditional is exactly the same as the one ending on line 282.

Comment (dra27) [reply Makefile:257]: Sorry - just looked at the raw file. GitHub (or my browser) is rendering the diff view badly - lines 253-256 are showing at the same indentation. Sorry for the noise...

Comment (dra27) [inline Makefile]: Sorry - just looked at the raw file. GitHub (or my browser) is rendering the diff view badly - lines 253-256 are showing at the same indentation. Sorry for the noise...

**Comment by gasche** [comment]
That's up to @whitequark preference, but my inclination would be "yes": this change fixes an user-observable issue -- so observable that the observers decided to fix the behavior. (It could go in "Bug fixes" or "Compiler distribution build system".)

**Comment by whitequark** [comment]
@gasche I have no real preference beyond having the bug fixed but your reasoning is convincing.

**Comment by dra27** [comment]
All looks good, thanks!
